### PR TITLE
Support using fp16 master weights and fp16/fp8 optimizer states in FusedAdam

### PR DESCRIPTION
**Pull request by kunlunl**
# Description

Add options to set the dtypes of master weights, exp_avg and exp_avg_sq of FusedAdam.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Support using fp32/fp16 master weights
- Support using fp32/fp16/fp8 exp_avg
- Support using fp32/fp16/fp8 exp_avg_sq

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by kunlunl** [comment]
@timmoon10 Hello, I noticed no one has commented on this MR for a long time, could you please take a look, or could you help find someone to review it?

File: tests/pytorch/test_fused_optimizer.py [hunk 1]
```diff
@@ -174,6 +175,223 @@ def test_frozen_model(self):
...
+                ref_params.append(p.detach().clone().float())
+                model_params.append(p)
+
+        options = {
+            "lr": 5e-4,
+            "betas": (0.9, 0.999),
+            "eps": 1e-08,
+            "weight_decay": 0,
+            "amsgrad": False,
+        }
```
Comment (timmoon10) [inline tests/pytorch/test_fused_optimizer.py:221]: The numerical tolerances are ~1e-3, so we should make sure that the optimizer has a detectable impact on the params:
```suggestion
        options = {
            "lr": 1,
            "betas": (0.1, 0.25),
            "eps": 1e-08,
            "weight_decay": 0,
            "amsgrad": False,
        }
```

**Review by timmoon10 — COMMENTED** [review]
Overall this looks good. It would be more general if we disentangled the state dtypes and state scaling (e.g. why not have scaled FP32 states or unscaled BF16 states?), but this does cover the cases in the [MS-AMP paper](https://arxiv.org/pdf/2310.18313).

For future reference, this PR adapts logic from https://github.com/NVIDIA/apex/pull/1771. This is a proof-of-concept with several opporunities for future improvement:
- TE kernel for computing absmax and scale
- Fusing scale/unscale within Adam kernel
- Reduce memory usage in optimizer step, perhaps by processing params in chunks
- Reduce memory usage in checkpointing, perhaps by storing checkpoint buffers in CPU

File: transformer_engine/pytorch/optimizers/fused_adam.py [hunk 2]
```diff
@@ -134,14 +168,211 @@ def __init__(
...
+
+        dtype = self.name_to_dtype_map[state_name]
+        if dtype == torch.uint8:
+            if state_name not in state:
+                state[state_name] = Float8Tensor.to_float8(torch.empty_like(param.data).float())
+                scale[state_name] = torch.ones([1], device=param.device)
+            self._apply_scale(state_name, unscaled_state, state[state_name], scale[state_name])
+        elif dtype == torch.half:
+            if state_name not in state:
+                state[state_name] = torch.empty_like(param.data).half()
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/pytorch/optimizers/fused_adam.py]: Done.

File: transformer_engine/pytorch/optimizers/fused_adam.py [hunk 3]
```diff
@@ -134,14 +168,211 @@ def __init__(
...
+        Arguments:
+            param (torch.nn.Parameter): One of parameters in this optimizer.
+            state_name (string): Name of optimizer states, can be one of 'exp_avg', 'exp_avg_sq',
+                and 'master_param`.
+            zero_buffer (bool): Whether to initialize the optimizer state with zeros.
+        """
+        buffer = torch.zeros_like(param).float() if zero_buffer else torch.empty_like(param).float()
+        dtype = self.name_to_dtype_map[state_name]
+        if dtype == torch.uint8:
+            self.state[param][state_name] = Float8Tensor.to_float8(buffer)
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/pytorch/optimizers/fused_adam.py]: Done.

File: transformer_engine/pytorch/optimizers/fused_adam.py [hunk 4]
```diff
@@ -134,14 +168,211 @@ def __init__(
...
+            param (torch.nn.Parameter): One of parameters in this optimizer.
+            state_name (string): Name of optimizer states, can be one of 'exp_avg', 'exp_avg_sq',
+                and 'master_param`.
+            unscaled_state (torch.Tensor): The original high-precision(FP32) state.
+        """
+        assert unscaled_state.dtype == torch.float32
+        state = self.state[param]
+        if param not in self._scales:
+            self._scales[param] = {}
+        scale = self._scales[param]
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/pytorch/optimizers/fused_adam.py]: Done.

File: transformer_engine/pytorch/optimizers/fused_adam.py [hunk 5]
```diff
@@ -68,11 +72,20 @@ class FusedAdam(torch.optim.Optimizer):
...
+        master_weights (bool, optional): whether to maintain FP32 master weights
+           in the optimizer with FP16/BF16 mixed precision training.
+            (default: False)
+        master_weight_dtype (torch.dtype, optional): The dtype of master weights.
+            If master_weights is False, this will be ignored.
+            (default: torch.float32)
+        exp_avg_dtype (torch.dtype, optional): The dtype of exp_avg.
+            (default: torch.float32)
+        exp_avg_sq_dtype (torch.dtype, optional): The dtype of exp_avg_sq.
+            (default: torch.float32)
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/pytorch/optimizers/fused_adam.py]: Added more comments.

File: transformer_engine/pytorch/optimizers/fused_adam.py [hunk 6]
```diff
@@ -112,9 +149,6 @@ def __init__(
         self.set_grad_none = set_grad_none
 
         self.capturable = capturable
-
-        if master_weights is not None:
-            assert isinstance(master_weights, list), "master_weights must be a list if provided"
         self.master_weights = master_weights
```
[the remark this answers is not in the document]
Comment (kunlunl) [reply transformer_engine/pytorch/optimizers/fused_adam.py:160]: Yes, I know this problem. I talked with @Wong4j offline and invited him to review this PR.
His MR in MCore (fuse dtype casting) has not been merged yet, so I put the "fusing dtype casting" function into a new MR in MCore, together with this precision-aware optimizer.

**Review by timmoon10 — APPROVED** [review]
LGTM, pending CI and confirmation from @Wong4j that this won't break Mcore integration.

**Comment by Wong4j** [comment]
LGTM. 
@timmoon10 This design is better. My mcore PR is not merged yet. So it won't break mcore.

**Pull request merged**